src/Gaffer/Reference.cpp

```diff
@@ -542,6 +542,10 @@ void Reference::loadInternal( const std::string &fileName )
 	{
 		PlugEdits::LoadingScope loadingScope( m_plugEdits.get() );
 		errors = script->executeFile( path.string(), this, /* continueOnError = */ true );
+		// set the childNodesAreReadOnly metadata again after loading is done
+		// this guarantees that the correct state is set, even if the exported file
+		// had the metadata baked in with a different value
+		MetadataAlgo::setChildNodesAreReadOnly( this, true, /* persistent = */ false );
```

Comment:
This is an issue right now at `IE`, because of how we can publish boxes from `DynamicBoxes`, where the boxes would necessarily have the "childNodesAreReadOnly" metadata set to `False` on the box, as that's necessary to make it editable. And then the `exportForReference()` would include that metadata.

We might want to also change the `exportForReference()` so that it ignores that metadata. @johnhaddon, let me know if you think that is also necessary, and if so, how do you think it should be implemented.

I did implement it at `IE` by simply removing the metadata temporarily while exporting the file. A likely better approach would be to do something on the serialiser itself, but I'm not sure how.

However, with this commit, changing the export is not strictly necessary.

Note that's the case only because we expect the metadata to be set at the very end of the reference file, so it won't impact `setNumericBookmarks()` included in the reference.

Reply:
As noted in the previous comment, I'd really like to avoid the proliferation of per-instance metadata, and to me this just feels like a workaround for a problem elsewhere.

> I did implement it at IE by simply removing the metadata temporarily while exporting the file. 

This also feels a little bit awkward. I wonder if it's an option for you to avoid per-instance metadata entirely? Could you register metadata against your box node type, using a lambda to query the state and return the appropriate value for `childNodesAreReadOnly` dynamically? Or if that's not an option, perhaps just use `persistent = False` for the per-instance metadata?

Either way, I imagine there may already be files in the wild with this metadata serialised in to them, and we will need a way of dealing with them? If so, perhaps we could use `Metadata::deregisterValue( "childNodesAreReadOnly" )` instead of `Metadata::setChildNodesAreReadOnly()`, so we're tidying up the stray metadata rather than adding to the problem.

Reply:
> This also feels a little bit awkward. I wonder if it's an option for you to avoid per-instance metadata entirely? Could you register metadata against your box node _type_, using a lambda to query the state and return the appropriate value for `childNodesAreReadOnly` dynamically? Or if that's not an option, perhaps just use `persistent = False` for the per-instance metadata?

The `childNodesAreReadOnly` is how I know the state of the box. It's the original source of the information, so it needs to be a per-instance metadata.

So, any changes to the system right now are likely to hit significant issues in terms of backward compatibility.

How big of a deal is this performance hit for using per-instance metadata? Would it be worth the effort of a major revamp? If I were to have the state somewhere else, would it be ok to use another per-instance metadata to store the state (and drive the lambda function you were suggesting), or would we have to use a plug value instead?

I likely could improve it, without a major overhaul, by only storing per-instance metadata for when we want the box to be editable (since the default state is supposed to be read-only). Right now, I use non-persistent per-instance metadata to store that read-only value. And since most boxes are in the read-only state, maybe that would be enough to avoid any performance issues regarding the use of per-instance metadata.

> 
> Either way, I imagine there may already be files in the wild with this metadata serialised in to them, and we will need a way of dealing with them? If so, perhaps we could use `Metadata::deregisterValue( "childNodesAreReadOnly" )` instead of `Metadata::setChildNodesAreReadOnly()`, so we're tidying up the stray metadata rather than adding to the problem.

Yes. If I can make it so that the default metadata for `References` is set to `True`, then I can just deregister them here.

I'll try that.

Reply:
> The childNodesAreReadOnly is how I know the state of the box. It's the original source of the information, so it needs to be a per-instance metadata.

Cool - I hadn't realised it was the primary source. In that case definitely don't worry about trying to switch to use some other source. What felt awkward was the remove-serialise-replace juggling in particular rather than the use of metadata. I was probably overstating the performance case against per-instance metadata a little, particularly now we've sorted out the signalling overhead. But we should definitely still use a type-based (rather than instance-based) registration in `Reference.cpp`, because conceptually we do want the same value for all instances of that type.

> Yes. If I can make it so that the default metadata for References is set to True, then I can just deregister them here.

Great - I think that's a good compromise.

Reply:
Addressed here: dceeb6dfaf2c5b4c3ca9ebb544e038b0ed4eedc7